Add `contextSummary` variable to AI system

colin-grant-work opened the pull request:
- Adds context to end of Coder agent's prompt
- Adds a function to allow agents to supplement the context

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does

Fixes #14895, Closes #14945

![image](https://github.com/user-attachments/assets/47fbbcfa-bca7-482a-a407-f36d9213ee4a)

<!-- Include relevant issues and describe how they are addressed. -->

#### How to test

1. Add a file to the context following the instructions in #14787
2. Then ask Coder about something in the context. The agent should know what you're talking about.
3. Ask Coder to add something to the context. It should succeed in doing so, and the file should pop up in the UI.

<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->



#### Follow-ups

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

1. I notice that files in the context are getting labeled with a leading slash, making them appear absolute. We could address that, if we think it's confusing.
2. I also notice that if you drag+drop a file from the explorer tree onto the AI input widget, the text input gets populated with the file information:

![image](https://github.com/user-attachments/assets/792da47f-ac67-4f25-bb63-2b39eb4d5208)


#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Attribution

<!-- If the changelog entry for this change should contain an attribution at the end (e.g. Contributed on behalf of x) add it in this section -->

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

JonasHelming commented:
Great, thank you for this integration!! The whole context story really significantly improves the work with Coder.

I did not look at the code yet, but tested a few things. My observations so far:

# Case 1: Context added in chat
This worked before and the additional context summary does nothing bad (as expected). 

# Case 2: Context in the chat + some files only in the context
When I add additional files without any further information, and it was not related to the task, the LLM did not even look at it. When I provided an additional file that was actually relevant for the task, the LLM looked at it, it probably guessed that from the file name and the context. Looks definitely good enough to merge this to me, but we have to evaluate in practise.

# Case 3: Context only in the context, not in the chat
It looked at the file(s) most of the time. Only if the prompt does not imply any implementation task, it did not. Example:
"Do not display a message to the user anymore about the tools" with context mcp-command-contribution.ts => It said: OK I won't :-D

So from case 2 and 3 we identify a **Potential issue 1**  that the LLM does not look at some files only in the context, but the user expects it to. I found only one prompt though, where this was really an issue. We should evaluate this in practice. If it is a problem, we can augment the prompt in iterations (during internal test) , e.g.:

- "Look at them if you need them for your task"
- "Always look at them to understand your task"

So issue 1 is not a blocker for this PR.


# Case 4: Agent builds the context by itself

**Issue 2**: If a file is added to the context multiple times, it appears there multiple times (@planger I believe this is an issue we should solve in the context layer, WDYT?)

**Issue 3** With my prompts, the LLM rather rarely "remembered" a file by itself.

I tried two options how to fix this:

Option A: Add the context add function to a sub bullet in the Context Retrieval section "If you find a file interesting for the task, remember it" => The issue here was that the LLM tended to add irrelevant files, but this needs to be evaluated.

Option B: Add the context function to a sub bullet in the apply change area "- **Remember changed file**: Use using~{context_addFile} to remember all files you change and files that were needed to create the changes." => Seemed to have better results than A as it does that in a phase, were it implements.

Option C: We add all modified files to the context in the modify functions ourselves (did not try). This very likely makes sense anyways.

Option A and B can be tried "on the go" as they are prompt modifications. Option C we have to implement.

My current option would be to go for Option C as a default. 
Reasoning:
- Our current agents do not tend to look at "other relevant files" by itself, e.g. for interface declarations. So the user has to typically provide these themselves and the LLM might add irrelevant things.
- We have the context add function in place and can evaluate it more in practice with a prompt variant.

All described issues IMHO do not block this PR, we can handle them in follow-ups

Thanks again for this great addition!

planger commented:
> Issue 2: If a file is added to the context multiple times, it appears there multiple times (@planger I believe this is an issue we should solve in the context layer, WDYT?)

Yes, this can be considered a bug. We shouldn't add the same element twice. I'll put it on my list.

JonasHelming commented:
A pretty 'satisfying' user prompt for me to test this was the following, works well with 40 and o3-mini. it is actually the follow-up of this PR, so the PR follow-ups itself :-)

---
@Coder Add a new variable to   #file:packages/ai-chat/src/common/context-summary-variable.ts that fully resolves the elements in the context including their content. Look at   #file:packages/ai-ide-agents/src/browser/context-functions.ts on how to resolve the context. Now add this variable to the prompt of the universal agent (which is define in   #file:packages/ai-ide-agents/src/common/universal-chat-agent.ts). similar to how it is used in the prompt of the coder agent.
---

Note that you should type the file variables out, as they are not added to the context when pasting (the prompt of course still works, but for the test this is important)

I literately adapted the system prompt to the one below. This worked pretty nicely, but it did not add files to the context by itself, so I still believe we should for for option C for now.

---
You are an AI assistant integrated into Theia IDE, designed to assist software developers with code tasks. You can interact with the code base and suggest changes.

## Context Retrieval
Use the following functions to interact with the workspace files if you require context:
- **~{getWorkspaceDirectoryStructure}**: Returns the complete directory structure.
- **~{getWorkspaceFileList}**: Lists files and directories in a specific directory.
- **~{getFileContent}**: Retrieves the content of a specific file.

## Propose Code Changes
To propose code changes or any file changes to the user, never print code or new file content in your response.

Instead, for each file you want to propose changes for:
  - **Always Retrieve Current Content**: Use getFileContent to get the latest content of the target file.
  - **Remember changed file**: Use ~{context_addFile} to remember all files you change and files that were needed to create the changes.
  - **Change Content**: Use ~{changeSet_writeChangeToFile} or ~{changeSet_replaceContentInFile} to suggest file changes to the user.  Only select and call one function per file.
  
## Additional Context

The following files have been provided for additional context. Some of them may also be referred to above. Always look at the relevant files to understand your task using getFileContent
{{contextSummary}}
---

JonasHelming commented:
Thought about it a bit more and I believe instead of option C, we should add a resolved version of the changeset (paths and state) to the prompt with another variable. This is now very easy to do, as we have the infrastructure in place. reasoning: the previously changed files are always interesting. I can add this in a follow-up PR.

Then we can go for option A with this PR and optimized from there.

@planger WDYT?

planger reviewed (commented):
Thank you very much! Code looks great to me and seems to work fine.

My main concern is how we make the context summary as interpretable as possible for the LLM and give the context variable provider more control over it (see my inline comment).

> Thought about it a bit more and I believe instead of option C, we should add a resolved version of the changeset (paths and state) to the prompt with another variable. This is now very easy to do, as we have the infrastructure in place. reasoning: the previously changed files are always interesting. I can add this in a follow-up PR.
> 
> Then we can go for option A with this PR and optimized from there.

I would agree. I think for now keeping those separated is simpler and more flexible. If we observe that it confuses the LLMs (which might be the case), we can reconsider. But we'll need more evaluation to confirm this even is a problem we would need to solve. :-)

JonasHelming commented:
Let's go with this prompt then: https://github.com/eclipse-theia/theia/pull/14971#issuecomment-2668098393

planger commented on packages/ai-chat/src/common/context-summary-variable.ts:
```diff
@@ -0,0 +1,46 @@
+
+export const CONTEXT_SUMMARY_VARIABLE: AIVariable = {
+    id: 'contextSummary',
+    description: nls.localize('theia/ai/core/todayVariable/description', 'Does something for today'),
```
I think this is a leftover from the previous implementation. :-)

JonasHelming commented:
for #14945  we should directly add a second function to this PR, something like the follewing (not compiling semi-pseudo code)

```
export const CONTEXT_RESOLVE_VARIABLE: AIVariable = {
    id: 'contextResolve',
    description: nls.localize('theia/ai/core/contextResolve/description', 'Resolves the full content of context elements'),
    name: 'contextResolve',
};

@injectable()
export class ContextResolvedVariableContribution implements AIVariableContribution, AIVariableResolver {
    registerVariables(service: AIVariableService): void {
        service.registerResolver(CONTEXT_RESOLVE_VARIABLE, this);
    }

    canResolve(request: AIVariableResolutionRequest, context: AIVariableContext): MaybePromise<number> {
        if (request.variable.name === CONTEXT_RESOLVE_VARIABLE.name) {
            return 50;
        }
        return 0;
    }

    async resolve(request: AIVariableResolutionRequest, context: AIVariableContext): Promise<ResolvedAIVariable | undefined> {
        if (!ChatSessionContext.is(context) || request.variable.name !== CONTEXT_RESOLVE_VARIABLE.name) { return undefined; }

        return {
            variable: CONTEXT_RESOLVE_VARIABLE,
            value: (() => {
                context.variables.map((contextElement: any) => {
                    const variable = context.session.context.getVariables().find(v => v.id === contextElement);
                    return variable ? `Content-ID: ${variable.arg}\n + Content: ${variable.contextValue}` : 'Variable not found.'
                }
                
            })()
        }
    };
}
```

JonasHelming commented:
@colin-grant-work Please test whether declining a changeste element and clearing the changeset works on this branch, I believe it did not when I tested, it works on master. But I am not sure

colin-grant-work commented on packages/ai-chat/src/common/chat-model.ts:
```diff
@@ -615,6 +640,42 @@ export class ChangeSetImpl implements ChangeSet {
     }
 }
 
+export class ChatContextManagerImpl implements ChatContextManager {
```
In this case, I was following the example of `ChangeSetImpl`, since that's the closest analogy as a small utility class to help the chat session, but I'm open to suggestions of more descriptive names. We have a few `Default...` classes, which might apply here.

planger replied:
Yeah, you are right, I'm happy with either naming and follow what you think feels best, given that we already have a few naming patterns in place here.

colin-grant-work commented on packages/ai-chat/src/common/context-summary-variable.ts:
```diff
@@ -0,0 +1,46 @@
+        if (!ChatSessionContext.is(context) || request.variable.name !== CONTEXT_SUMMARY_VARIABLE.name) { return undefined; }
+        return {
+            variable: CONTEXT_SUMMARY_VARIABLE,
+            value: context.session.context.getVariables().filter(variable => !!variable.arg).map(variable => `- ${variable.arg}`).join('\n')
```
@planger, thanks for the suggestion. While I like the idea, and, I've basically implemented it for the new `contextDetails` variable for use in agents other than Coder, it has the downside that the `value` field isn't available unless the variable has been resolved, and I don't think we have access to resolved variables at this point - after all, we're busy resolving one ourselves. Or is there somewhere where resolved values are cached (and would it be safe to use that cache)?

We could use the static `description` value associated with the variable itself, something like the file variable's `'Resolves the contents of a file'`. That's not really directed at the LLM, but we could add a field that would be.

planger replied:
Please see my suggestion on the variable dependencies.

Regarding the description, I totally agree. This was a misconception on my end. The description is not targeted at the LLM and therefore is probably in most cases not a great fit.

colin-grant-work commented:
@planger, @JonasHelming, I've pushed a new version of the code that addresses some of the comments above, but there is likely still some work to do re: formatting things as intelligibly as possible.

 - I have added two new variables, one for change sets (added to Coder's system prompt), and one for context details, added to the universal agent's prompt.
 - Re: context details, @planger proposed adding various details of resolved variables to the `contextSummary` variable. Since that variable is intended for use with Coder, which can resolve some details on its own, that seems perhaps overkill, but I have attempted to resolve the variables in context for the `contextDetails` variable. Let me know what you think about that approach, but also let me know if there's some way to access resolved variables while resolving other variables; then we could use them as we see fit in both cases, and not resolve them ourselves, as I've done for `contextDetails`.
 - Since there's a lot of data for `contextDetails`, I've just JSON stringified what's available - maybe there's a better way? Maybe I should do that for the change set, as well?

@JonasHelming, I have fixed a bug re: showing the deletion of change sets, so that should be working. On the other hand, I'm still not seeing the agent respond to changes in context during the session, while it seemed like in your interactions, the agent was resolving changes to context later in the session, and not just for the initial interaction.

planger reviewed (commented):
Thank you very much for this iteration! I think this goes into the right direction.

I didn't consider the variable dependency issue before (sorry for that), but I feel this is a valid general use case. So I suggest to explicitly support dependent variables.

We need to ensure that we
* Don't inject the variable service into variable resolvers (cyclic DI)
* Aim at taking care of cyclic variable dependencies
* Avoid duplicate resolution

We might be able to address those constraints by handing in a `resolveDependency` callback to resolvers and enhance the variable resolution method.

### Extend the variable resolver

```ts
export interface AIVariableResolverWithVariableDependencies extends AIVariableResolver {
    resolve(
        request: AIVariableResolutionRequest,
        context: AIVariableContext,
        resolveDependency: (req: AIVariableResolutionRequest) => Promise<ResolvedAIVariable | undefined>
    ): Promise<ResolvedAIVariable | undefined>;
}

function isResolverWithDependencies(resolver: AIVariableResolver | undefined): resolver is AIVariableResolverWithVariableDependencies {
    return resolver !== undefined && resolver.resolve.length >= 3;
}
```

### Enhance the resolveVariable method of the variable service

```ts
interface CacheEntry {
    promise: Promise<ResolvedAIVariable | undefined>;
    inProgress: boolean;
}

export class DefaultAIVariableService implements AIVariableService {
    // ... existing properties and methods

    async resolveVariable(
        request: AIVariableArg,
        context: AIVariableContext,
        cache: Map<string, CacheEntry> = new Map()
    ): Promise<ResolvedAIVariable | undefined> {
        const variableName = typeof request === 'string'
            ? request
            : typeof request.variable === 'string'
                ? request.variable
                : request.variable.name;

        if (cache.has(variableName)) {
            const entry = cache.get(variableName)!;
            if (entry.inProgress) {
                this.logger.warn(`Cycle detected for variable: ${variableName}. Skipping resolution.`);
                return undefined;
            }
            return entry.promise;
        }

        const entry: CacheEntry = { promise: Promise.resolve(undefined), inProgress: true };
        cache.set(variableName, entry);

        const promise = (async () => {
            const variable = this.getVariable(variableName);
            if (!variable) {
                return undefined;
            }
            const arg = typeof request === 'string' ? undefined : request.arg;
            const resolver = await this.getResolver(variableName, arg, context);
            let resolved: ResolvedAIVariable | undefined;
            if (isResolverWithDependencies(resolver)) {
                resolved = await resolver.resolve(
                    { variable, arg },
                    context,
                    async (depRequest: AIVariableResolutionRequest) =>
                        this.resolveVariable(depRequest, context, cache)
                );
            } else {
                resolved = await resolver?.resolve({ variable, arg }, context);
            }
            return resolved ? { ...resolved, arg } : undefined;
        })();

        entry.promise = promise;
        promise.finally(() => {
            entry.inProgress = false;
        });

        return promise;
    }
}
```

### Summary

- **AIVariableResolverWithVariableDependencies** lets a resolver access dependencies through a callback during its resolution.
- The updated `resolveVariable` method uses a promise-based cache with an `inProgress` flag for cycle detection. This way, if a variable is already being resolved, it’s skipped—avoiding cycles.
- With this approach, resolved variables are accessible during the resolution of dependent ones. This could be leveraged for both contextSummary (letting Coder resolve details) and contextDetails, reducing redundant resolution.

What do you think about this approach?

Maybe I'm over-complicating this, but I feel like letting variables access other variables might be a useful thing in general, and if it is, we should explicitly support it.

If you or others disagree that we should support that, I'm happily follow your lead and think your approach is fine, given it is rather local and puts the responsibility of using this approach carefully into the hands of a specific variable resolver.

planger commented on packages/ai-chat/src/common/change-set-variable.ts:
```diff
@@ -0,0 +1,64 @@
+};
+
+@injectable()
+export class ContextSummaryVariableContribution implements AIVariableContribution, AIVariableResolver {
```
I think we should name this `ChangeSetSummaryVariableContribution` right?

planger replied:
Currently this variable is targeted at files, even though the context may contain arbitrary context variables, like symbols or even domain-specific things.

If we want to keep it that way, we should also reflect that in the naming. It is already reflected in the description.

In general, I tend to believe we can generalize that, because the only file specific thing is currently the description and the word "file" in line 60. Both can be avoided by using the `contextSummary` field or the variable name.

planger commented on packages/ai-chat/src/common/context-details-variable.ts:
```diff
@@ -0,0 +1,61 @@
+        const data = values.filter((candidate): candidate is ResolvedAIVariable => !!candidate)
+            .map(resolved => ({
+                variableKind: resolved.variable.name,
+                variableKindDescription: resolved.variable.description,
```
You are probably right to remove this.

colin-grant-work commented:
@planger, following our discussion, my latest push mostly cleans up the existing implementation.

 - I've corrected the naming of and bound the ContextDetails, ContextSummary, and ChangeSet variable contributions.
 - I've  removed the context summary variable's focus on files and exposed the context value resolution function to the both the Architect and Coder agents so that they can context items generically.
 - I've exposed the ContextDetails 
 - I've updated the various places we pass context into variable resolvers to consistently use the `ChatSessionContext` interface where available.

There is one outstanding question from https://github.com/eclipse-theia/theia/issues/14945, namely whether we want to provide more details than the file names and status and provide e.g. the full text of a given suggestion.

Re: follow ups, I have not implemented your suggestions re: explicit support for dependent variables. In the event, it wasn't necessary because we already pre-resolve the context, with which the variables in this PR are concerned. In my view, we shouldn't eagerly resolve the context, because we don't know that it will actually be used, but we should implement some kind of variable caching (perhaps on the request level?) and better support for variable dependencies, since if we don't eagerly resolve context variables, then the context summary will definitely fall into that category.

You'll also see that I've changed the passing of the context a bit: rather than having it passed in as an argument to the chat service, I've located it on the ChatModel. That makes it easier to pass around as part of variable resolution - e.g. when handling system messages - but if you believe that it's important that the context be separated from the model, I'm open to discussion.

JonasHelming commented:
Note: we just decided in a direct discussion to split the generic context summary variable provided by the framework and the one used by Coder and Architect. the reasoning is that we want to make it maximum simple for Coder and Architect and as the can already access files, they just need the path. Furthermore, as they are part of the IDE, we have under control, which additional future context element types they have to deal with and can handle this explicitly. Having the generic variable in the framework provides value for adopters of Theia AI, as they can directly use this as a good default.

JonasHelming commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
+  
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above.\
```
```suggestion
The following files have been provided for additional context. Some of them may also be referred to above. \
```

JonasHelming commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
-  - **Change Content**: Use ~{changeSet_writeChangeToFile}${withSearchAndReplace ? ' or ~{changeSet_replaceContentInFile}' : ''} to suggest file changes to the user.\
-  ${withSearchAndReplace ? 'Only select and call one function per file.' : ''}`,
+- **Always Retrieve Current Content**: Use ${FILE_CONTENT_FUNCTION_ID} to get the latest content of the target file.
+- **Change Content**: Use ~{changeSet_writeChangeToFile}${withSearchAndReplace ? ' or ~{changeSet_replaceContentInFile}' : ''} to suggest file changes to the user.\
```
```suggestion
- **Change Content**: Use ~{changeSet_writeChangeToFile}${withSearchAndReplace ? ' or ~{changeSet_replaceContentInFile}' : ''} to propose file changes to the user.\
```

JonasHelming commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
+{{${CONTEXT_FILES_VARIABLE_ID}}}
+
+## Previously Proposed Changes
+Changes have been proposed for the following files. Some suggestions may have been accepted, while others may still be pending.
```
```suggestion
You have previously proposed changes for the following files. Some suggestions may have been accepted by the user, while others may still be pending.
```

JonasHelming commented on packages/ai-ide/src/common/architect-prompt-template.ts:
```diff
@@ -38,5 +39,10 @@ Use the following functions to interact with the workspace files as needed:
 
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above.\
```
```suggestion
The following files have been provided for additional context. Some of them may also be referred to above. \
```

JonasHelming reviewed (commented):
I tested Coder, Architect and Universal and did not find any case, where this does not work well so far. I added some minor suggestions.
I will let Philip apporve this, if the framework part is fine.

Great work, thank you

planger commented on packages/ai-chat/src/common/context-details-variable.ts:
```diff
@@ -0,0 +1,53 @@
+
+export const CONTEXT_DETAILS_VARIABLE: AIVariable = {
+    id: CHAT_CONTEXT_DETAILS_VARIABLE_ID,
+    description: nls.localize('theia/ai/core/contextDetailsVariable/description', 'Provides full text values and descriptions for all context variables.'),
```
```suggestion
    description: nls.localize('theia/ai/core/contextDetailsVariable/description', 'Provides full text values and descriptions for all context elements.'),
```

planger commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
+
+The following files have been provided for additional context. Some of them may also be referred to above. \
+Always look at the relevant files to understand your task using getFileContent
+{{${CONTEXT_FILES_VARIABLE_ID}}}
```
Probably not a problem, but we imply in the text above that we always have files, even though there might not be any context files, which will lead to an empty value here. We might want to change the phrasing a bit (like below) and use "if any, the files below...", but I'm not sure if that would impact the quality of the LLM response. Anyway, this can be evaluated in practice and changed later on.

JonasHelming replied:
AFAIK, {{${CONTEXT_FILES_VARIABLE_ID}}} resolves to "no files" or something similar if undefned, so should be fine

planger replied:
No, it returns empty string then:
https://github.com/eclipse-theia/theia/pull/14971/files#diff-0a238071acebd16667ceb34ff1da44e3704ad463f083e9bc47383371cec1bf34R43

JonasHelming replied:
I think I leave this for now to avoid "if". My assumption is that an empty list is easier to comprehend than another "if" in the sentence

planger commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
+{{${CONTEXT_FILES_VARIABLE_ID}}}
+
+## Previously Proposed Changes
+You have previously proposed changes for the following files. Some suggestions may have been accepted by the user, while others may still be pending.
```
Same as above, we imply that there always is a changeset, even though there might not be any. But as above, this can be evaluated in practice and changed later on.

JonasHelming replied:
agreed, did not have issues with this so far

planger reviewed (approved):
Thank you very much @colin-grant-work for this excellent work and for pushing through all of the discussions we had on this PR!

This looks great and works very well. I didn't test in detail as @JonasHelming already did, but what I tested seemed pretty good.

I don't have any comment that should block merging this change, just a few details that may well be first evaluated in practice or that can be done in a follow-up. Besides my inline comments, I'd suggest to rename and move:
* `ChangeSetVariableContribution`
* `ContextSummaryVariableContribution`
Both are somewhat file-specific at the moment, which is fine as discussed---we can generalize them later, but then we should probably move them to `ai-ide` and rename them. This would then also be in line with e.g. the `ContextFilesVariableContribution`.

Even that however, can be done in a follow-up after the pre-release as it doesn't impact the behavior at all. So I'm fine to merge this and we'll create follow-ups.

JonasHelming commented on packages/ai-ide/src/common/architect-prompt-template.ts:
```diff
@@ -38,5 +39,10 @@ Use the following functions to interact with the workspace files as needed:
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above. \
+Always look at the relevant files to understand your task using getFileContent
```
The function is already referenced above. Did you test this change that mentioning the function twice does not introduce a regression? otherwise I can do.

planger replied:
Good point! I didn't test this on this PR, but already referenced functions multiple times before in other cases without a problem. A few things have changed in the meantime, so it is probably worth re-testing. If it really is causing an issue, we should definitely track a bug report on that :-)

JonasHelming commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above. \
+Always look at the relevant files to understand your task using getFileContent
```
same question

planger replied:
https://github.com/eclipse-theia/theia/pull/14971#discussion_r1965064048

JonasHelming commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
+  
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above. \
```
same comment as above

planger commented on packages/ai-ide/src/common/architect-prompt-template.ts:
```diff
@@ -38,5 +39,10 @@ Use the following functions to interact with the workspace files as needed:
 
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above. \
```
Yes, that's even better

JonasHelming commented on packages/ai-ide/src/common/architect-prompt-template.ts:
```diff
@@ -38,5 +39,10 @@ Use the following functions to interact with the workspace files as needed:
 
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above. \
```
```suggestion
The following files have been provided for additional context. Some of them may also be referred to by the user. \
```

JonasHelming commented on packages/ai-ide/src/common/coder-replace-prompt-template.ts:
```diff
@@ -36,14 +39,26 @@ Use the following functions to interact with the workspace files if you require
+  
+## Additional Context
+
+The following files have been provided for additional context. Some of them may also be referred to above. \
```
```suggestion
The following files have been provided for additional context. Some of them may also be referred to by the user. \
```

JonasHelming reviewed (approved):
Tested again with the minor updates, no issues found
